storage/range_command.go

```diff
@@ -346,16 +352,39 @@ func (r *Range) InternalRangeLookup(batch engine.Engine, args *proto.InternalRan
 	// MaxRanges.
 	metaPrefix := proto.Key(args.Key[:len(engine.KeyMeta1Prefix)])
 	nextKey := proto.Key(args.Key).Next()
-	// Always false, at least when called from the DistSender.
-	consistent := args.ReadConsistency != proto.INCONSISTENT
-	kvs, err := engine.MVCCScan(batch, nextKey, metaPrefix.PrefixEnd(), rangeCount, args.Timestamp, consistent, args.Txn)
+	kvs, err := engine.MVCCScan(batch, nextKey, metaPrefix.PrefixEnd(), rangeCount, args.Timestamp, false, args.Txn)
 	if err != nil {
-		reply.SetGoError(err)
-		return
+		if wiErr, ok := err.(*proto.WriteIntentError); ok && args.IgnoreIntents {
+			// NOTE (subtle): in general, we want to try to clean up dangling
+			// intents on meta records. However, if we're in the process of
+			// cleaning up a dangling intent on a meta record by pushing the
+			// transaction, we don't want to create an infinite loop:
+			//
+			// intent! -> push-txn -> range-lookup -> intent! -> etc...
+			//
+			// Instead we want:
+			//
+			// intent! -> push-txn -> range-lookup -> ignore intent, return old/new ranges
+			//
+			// On the range-lookup from a push transaction, we want to be
+			// insensitive to intents and avoid resolving them. However, because
+			// the range may or may not have changed depending on the success of
+			// the split/merge/update, we return the range info without intent
+			// errors. However, because we don't know whether the range update
+			// succeeded or failed, we are forced to return old or new randomly.
```

Comment:
I find this comment unclear; how about:

On the range-lookup from a push transaction, we therefore want to suppress WriteIntentErrors and return a value anyway. But which value? We don't know whether the range update succeeded or failed, but if we don't return the correct range descriptor we may not be able to find the transaction to push. Since we cannot know the correct answer, we choose randomly between the pre- and post- transaction values. If we guess wrong, the client will try again and get the other value (within a few tries).